Framework/DataHandling/src/SinglePeriodLoadMuonStrategy.cpp

```diff
@@ -0,0 +1,185 @@
+#include "MantidAPI/Algorithm.h"
+#include "MantidAPI/Axis.h"
+#include "MantidAPI/GroupingLoader.h"
+#include "MantidAPI/WorkspaceFactory.h"
+#include "MantidDataHandling/ISISRunLogs.h"
+#include "MantidDataHandling/LoadMuonLog.h"
+#include "MantidDataHandling/LoadMuonNexusV2Helper.h"
+#include "MantidKernel/UnitFactory.h"
+#include "MantidKernel/UnitLabelTypes.h"
+
+namespace Mantid {
+namespace DataHandling {
+
+using namespace API;
+using namespace NeXus;
+using namespace HistogramData;
+using std::size_t;
+using namespace DataObjects;
+
+// Constructor
+SinglePeriodLoadMuonStrategy::SinglePeriodLoadMuonStrategy(
+    Kernel::Logger &g_log, const std::string &filename, NXEntry &entry,
+    Workspace2D_sptr workspace, int entryNumber, bool isFileMultiPeriod)
+    : LoadMuonStrategy(g_log, filename), m_entry(entry), m_workspace(workspace),
+      m_entryNumber(entryNumber), m_isFileMultiPeriod(isFileMultiPeriod),
+      m_detectors(getLoadedDetectors()) {}
+
+/**
+ * Loads Muon specific logs into the stored workspace
+ * These are logs which are not loaded by LoadISISNexus
+ */
+void SinglePeriodLoadMuonStrategy::loadMuonLogData() {
+
+  auto &run = m_workspace->mutableRun();
+  auto runSample = m_entry.openNXGroup("sample");
+
+  if (runSample.containsDataSet("temperature")) {
+    float temperature = runSample.getFloat("temperature");
+    run.addProperty("sample_temp", static_cast<double>(temperature));
+  }
+
+  if (runSample.containsDataSet("magnetic_field")) {
+    float magn_field = runSample.getFloat("magnetic_field");
+    run.addProperty("sample_magn_field", static_cast<double>(magn_field));
+  }
+  std::string mainFieldDirection =
+      LoadMuonNexusV2Helper::loadMainFieldDirectionFromNexus(m_entry);
+  run.addProperty("main_field_direction", mainFieldDirection);
+
+  double firstGoodData =
+      LoadMuonNexusV2Helper::loadFirstGoodDataFromNexus(m_entry);
+  run.addProperty("FirstGoodData", firstGoodData);
+}
+
+/**
+ * Loads the good frames data into the stored workspace object
+ */
+void SinglePeriodLoadMuonStrategy::loadGoodFrames() {
+  // Overwrite existing log entry
+  auto &run = m_workspace->mutableRun();
+  run.removeProperty("goodfrm");
+
+  NXInt goodframes = LoadMuonNexusV2Helper::loadGoodFramesDataFromNexus(
+      m_entry, m_isFileMultiPeriod);
+
+  if (m_isFileMultiPeriod) {
+    run.addProperty("goodfrm", goodframes[static_cast<int>(m_entryNumber - 1)]);
+  } else {
+    run.addProperty("goodfrm", goodframes[0]);
+  }
+}
+
+/**
+ * Loads detector grouping.
+ * If no entry in NeXus file for grouping, load it from the IDF.
+ * stored
+ * @returns :: Grouping table
+ */
+Workspace_sptr SinglePeriodLoadMuonStrategy::loadDetectorGrouping() const {
+
+  auto grouping = LoadMuonNexusV2Helper::loadDetectorGroupingFromNexus(
+      m_entry, m_detectors, m_isFileMultiPeriod);
+  DataObjects::TableWorkspace_sptr table =
+      createDetectorGroupingTable(m_detectors, grouping);
+
+  Workspace_sptr table_workspace;
```

Comment:
Do you need this extra variable? You should be able to `return table` and the pointer will be upcast automatically.